File: 1679._Max_Number_of_K-Sum_Pairs/pairSumEqualToK.py
def pairSumEqualToK(arr, k):
    count = 0
    hash = {}
    for i in range(len(arr)):
        if k - arr[i] in hash and hash[k - arr[i]] > 0:
            count += 1
            hash[k - arr[i]] -= 1
        else:
         if arr[i] in hash and hash[arr[i]] > 0:
            hash[arr[i]] += 1
         else:
            hash[arr[i]] = 1
    return count
    
print(pairSumEqualToK([1,2,3,4], 5))

File: 1679._Max_Number_of_K-Sum_Pairs/pairSumEqualToK.js
const pairSumEqualToK = (arr, k) => {
    const hash = {};
    let noOfPairs = 0;
    for (let i = 0; i < arr.length; i++) {
        if (hash[k - arr[i]]) {
            noOfPairs++;
            hash[k - arr[i]]--;
        } else {
            !hash[arr[i]] ? hash[arr[i]] = 1 : hash[arr[i]]++;
        }
    }
    return noOfPairs;
}
console.log("pairSumEqualToK", pairSumEqualToK([1,2,3,4], 5));